django/queryset/README.md

```diff
@@ -0,0 +1,340 @@
+# QuerySet
+
+## Introduction
+
+This guideline will base on these models:
+
+```python
+class User(models.Model):
+    username = models.CharField(max_length=50)
+
+
+class Person(models.Model):
+    user = models.ForeignKey(
+        'User',
+        on_delete=models.CASCADE,
+        related_name='people',
+    )
+    first_name = models.CharField(max_length=50)
+    last_name = models.CharField(max_length=150)
+    employee_id = models.CharField(max_length=15, unique=True)
+    active = models.Boolean(default=True, blank=True)
+
+
+class Department(models.Model):
+    name = models.CharField(max_length=255)
+
+    members = models.ManyToManyField(
+        'Person',
+        related_name='departments',
+    )
+```
+
+## Filtering
+
+### Joint Field Filtering
+
+```python
+# Yes
+def find_people_yes(query: str):
+    return Person.objects.annotate(
+        full_name=Concat('first_name', Value(' '), 'last_name'),
+    ).filter(
+        full_name__icontains=query,
+    )
+
+# No
+def find_people_no(query: str):
+    return Person.objects.filter(
+        Q(first_name__icontains=query)
+        | Q(last_name__icontains=query)
+    )
+```
+
+First, we try to filter people with name `John`...
+
+```python
+people = find_people_yes(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+
+people = find_people_no(query='John')
+# return: <QuerySet [<Person: John Doe>, <Person: John Wick>]>
+```
+
+Everything seems fine, right? Both techniques return the same results. But what about filtering person's name by `John Doe`?
+
+```python
+people = find_people_yes(query='John Doe')
+# return: <QuerySet [<Person: John Doe>]>
+
+people = find_people_no(query='John Doe')
+# return: <QuerySet []>
+```
+
+Now you see the difference? It's because the second method (`find_people_no`) does not support filtering across two fields.
+
+So if you have a requirement like this, you should join two fields together before filter them.
+
+### Using `.count()` instead of `len()`
+
+```python
+# Yes
+Person.objects.count()     # SELECT COUNT(*) FROM person
+
+# No
+len(Person.objects.all())  # SELECT * FROM person
+```
+
+Selecting aggregate function from database always consume less time that getting every records.
+
+### Using `.exists()` to check if any data exists
+
+```python
+# Yes
+if Person.objects.filter(first_name='John').exists():
+    do_something()
+
+# No
+if Person.objects.filter(first_name='John'):
+    do_something()
+```
+
+Using `.exists()` consume less resources than not using it.
+
+### Using `.first()` instead of calling `.exists()`, followed by `.get()`
+
+```python
+# Yes
+matched_person = Person.objects.filter(employee_id='123').first()
+if matched_person:
```

Comment:
```suggestion
if matched_person is not None:
```
Prefer `is not None` for null check krub

Reply:
Ah, roger krub.